Support for recursive config data structures

vigoo opened the pull request:
Support for recursive structures (https://github.com/zio/zio-config/issues/407)

jdegoes commented:
@vigoo As an alternative, could you make any binary or unary operator lazy in both operands? 

The main challenge would be making sure the methods don't blow up on recursive structures.

vigoo commented:
I think making them lazy is not enough. When reading a recursive structure, the current implementation does the following, assuming that the recursion is done by a `nested(l)(rec).optional`:
- reach `Optional`, eval inside
- reach `Nested`, it just changes the context by adding a new path segment to the current path, but otherwise does nothing
- the next level is most likely `Zip` (or eventually gets into a `Zip` because otherwise it's a recursive structure without holding any additional data than itself). Here it evaluates both the left and the right side, and gets into an endless recursion if either of its operands are the recursive nesting.

My first attempt to solve it was trying to break this by modifying the `Zip` behavior in this special case, but it's quite specific and only works if the nesting is at the right side, and it's rather ugly that the termination depends on the failure of the `Zip`s _other_ side.

The real issue is that the `Nested` step does not check the existence of the tree node it steps into. Now that as I wrote this down I realized it actually could do that, and that together with making the `ConfigDescriptor`s lazy would solve the issue completely. 
I'll try it out soon.

vigoo commented:
@jdegoes see the new approach, I think it's much better :) If it's OK for you too, I will check (write some tests) if config writing and doc generation is recursion-safe too and then it should be ready.

vigoo commented:
Fixed most of the issues caused by this change. One thing that broke and I decided changing test assertions is that it no longer reports individual elements of a nested record if the whole record is missing from the property tree (so instead of `List(missing a.x, missing a.y)` it's going to be `List(missing a)`). I think this is expected, but if necessary it would be possible to gather all the leaves missing from inside. 

Remaining:
- In case of `nested("x")(_ ?? "desc")` "desc" is not reported on the missing "x" because it's checked one lever outer. I fixed this in the Magnolia derive (by rearranging `Describe`s) but then realized that it should be fixed in a more general way, by looking ahead for descriptions in `ReadModule` on nesting failure, _or_ by moving out the inner description at the time of constructing the nested descriptor. 
- `handleDefaultValues` is broken because it can no longer decide the `baseConditionForFallBack` just based on the numeber of errors vs number of things in the subtree. This is because of how the `MissingValue` is now reported for the whole subtree not the individual elements as I described above.
- `complex source: read(descriptor from typesafeSource) == read(descriptor from write(descriptor, read(descriptor from typesafeSource)))` is also broken, did not investigate it yet, may be related to the other issues
- `shapeless` tests are broken, most likely similar change required as I did in the magnolia derivation
- Test recursive structures with the write module and the documentation generator module

I will address these tomorrow.

vigoo commented:
Note: the issue surfaced by the `TypesafeConfigReadWriteTest` test was a more general one. It came up in that test because I changed the magnolia derivation to move the optional out of nested (as nested now checks existence); This lead to the following case in  `b[2].c[0].d[3]`:

`left: Record(e -> Sequence(Nil), z -> Sequence(Nil))
 right: Empty`
where the `Empty` is the absence of `y`.

The `merge` on `PropertyTree` merged these two to `Empty` because it treated the record containing only empty sequences empty. I fixed this and wrote a small test in `ReadWriteRoundtripTest` that checks this behavior by writing/reading this:

```scala
val config = (list("a")(string) |@| nested("b")(string).optional).tupled
val data = (Nil, None)
```

vigoo commented:
I removed one test ("Absence of an optional product and a required field within another optional product returns none") as in my opinion it should not be true, and with the changes it broke. Please let me know if you disagree and it should be fixed instead.

It was accepting this:
```
detail: {
  a : 10
  b : {}
}
```
as `b = None` even though it misses a required field. I think reporting the missing field in this case is correct, the `None` case is expressed by `b = null` or a missing `b`.

afsalthaj commented on core/src/main/scala/zio/config/ConfigDescriptorModule.scala:
```diff
@@ -202,40 +202,93 @@ trait ConfigDescriptorModule extends ConfigSourceModule { module =>
+          go(config.get(), prefix)
+      }
+    go(config, "")
+    builder.toString()
```
Is this function necessary?

afsalthaj commented:
There is a few more details that I need to review. I will get this done over the weekends. I can see explanation on a removed test case. I will need to get my head around that as well.

vigoo commented on core/src/main/scala/zio/config/ConfigDescriptorModule.scala:
```diff
@@ -202,40 +202,93 @@ trait ConfigDescriptorModule extends ConfigSourceModule { module =>
+          go(config.get(), prefix)
+      }
+    go(config, "")
+    builder.toString()
```
I did not want to leave it in the code first, but then had to re-add several times while debugging and thought that it may help people in future development. Now that the descriptors are lazy, without this it was very inconvenient to examine a descriptor subtree in debugger / in debug logs.
Should I remove it?

vigoo commented on core/src/main/scala/zio/config/ConfigDescriptorModule.scala:
```diff
@@ -202,40 +202,93 @@ trait ConfigDescriptorModule extends ConfigSourceModule { module =>
 
+  case class LazyConfigDescriptor[A](get: () => ConfigDescriptor[A]) {
+    def map[B](f: ConfigDescriptor[A] => ConfigDescriptor[B]): LazyConfigDescriptor[B] =
+      LazyConfigDescriptor(() => f(get()))
```
I intentionally did not do that, so it's more visible everywhere that `get()` does evaluate the subtree. 
But if you'd prefer to hide it I'll add a `.value`

vigoo commented on core/src/main/scala/zio/config/ConfigDocsModule.scala:
```diff
@@ -68,20 +85,14 @@ trait ConfigDocsModule extends WriteModule {
+            val desc = filterDescriptions(descriptionsUsedAlready, descriptions)
 
             TableRow(previousPaths, Some(Table.Format.Primitive), desc, None, sources.map(_.name)).asTable
+          case ConfigDocs.Recursion(sources) =>
```
My idea was that you either reach a `Leaf` or a `Recursion`.
If it would be `Recursive`, then (in my head) it would more like `Nested`, and I would expect it to not a leaf.

vigoo commented on core/src/main/scala/zio/config/ConfigDocsModule.scala:
```diff
@@ -371,6 +382,7 @@ trait ConfigDocsModule extends WriteModule {
           case Format.Nested        => "nested"
           case Format.AnyOneOf      => "any-one-of"
           case Format.AllOf         => "all-of"
+          case Format.Recursion     => "recursion"
```
Good question :) Probably we should render a real example and see how it feels

vigoo commented on core/src/main/scala/zio/config/PropertyTree.scala:
```diff
@@ -62,7 +62,7 @@ sealed trait PropertyTree[+K, +V] { self =>
     case Empty           => true
     case Leaf(_)         => false
-    case Record(value)   => value.values.forall(_.isEmpty)
+    case Record(value)   => value.values.isEmpty
```
I found this issue with the existing `TypesafeConfigReadWriteTest`, and added a much smaller test case too to demonstrate it, it's `empty sequence zipped with optional nested` in `ReadWriteRoundtripTest`. 
Basically the original code was not able to write and re-read `(Nil, None)` via `(list("a")(string) |@| nested("b")(string).optional).tupled`

afsalthaj replied:
Fantastic! Good catch.

vigoo commented on core/src/main/scala/zio/config/ReadModule.scala:
```diff
@@ -268,19 +310,38 @@ private[config] trait ReadModule extends ConfigDescriptorModule {
+        case ConfigDescriptorAdt.Zip(left, right) =>
+          countZipSize(left.get()) + countZipSize(right.get())
+        case _ => 1
+      }
```
It was not straightforward how to change the logic in `handleDefaultValues`, but then I came up with this implementation. I'll try to reason about it.

The change is required because now `nested` terminates with an error if the subtree does not exists, while before it was just moving the context inner, and then each `Source` produced a separate error.

This function is only used in the following condition:
```
val baseConditionForFallBack =
      hasOnlyMissingValuesAndZeroIrrecoverable && sizeOfZipAndOrErrors(error) == requiredZipAndOrFields(config)
```
which basically tests that "everything is missing", in which case we replace the whole thing with a default value.

I changed the `Zip`, `OrElse` and `OrElseEither` cases. The idea is that each element of these is either:
- a `Source` in which case it counts as `1`
- a `Nested` in which case if the whole thing is missing, it terminates, so it counts as `1`. If it was only partially missing, it would not be `1` but then we have a count mismatch and the fallback condition is false which is good.
- Something else like `XmapEither`, or `Describe`, etc that does not affect the count

Because of this, I think that these "count a flat size" implementations provide the expected result for the above fallback condition (but only usable for that).

I can rewrite it in a more generic form as you suggested.

afsalthaj replied:
👍

afsalthaj commented on typesafe-magnolia-tests/src/test/scala/zio/config/typesafe/TypesafeConfigOptionalTest.scala:
```diff
@@ -782,14 +759,8 @@ object TypesafeConfigOptionalTest
-                  MissingValue(List(Key("a"), Key("a"), Key("b1")), List("optional value", "value of type string")),
-                  MissingValue(List(Key("a"), Key("a"), Key("a1")), List("optional value", "value of type string"))
+                  MissingValue(List(Key("a"), Key("c")), List("optional value")),
+                  MissingValue(List(Key("a"), Key("a")), List("optional value"))
```
This is good in the sense that the `optional` value is tagged at the root level instead of leaves as in the current version.

I am not entirely sure if its a good idea to avoid telling the user the entire tree that was missing instead of giving a hint that `a.c` and `a.a` is missing.

```scala

Right(ReadError:
╥
╠══╦══╗
║  ║  ║
║  ║  ╠══╦══╗
║  ║  ║  ║  ║
║  ║  ║  ║  ╠─MissingValue
║  ║  ║  ║  ║ path: a.c.b1
║  ║  ║  ║  ║ Details: optional value, value of type string
║  ║  ║  ║  ▼
║  ║  ║  ║
║  ║  ║  ╠─MissingValue
║  ║  ║  ║ path: a.c.a1
║  ║  ║  ║ Details: optional value, value of type string
║  ║  ║  ▼
║  ║  ║
║  ║  ╠══╦══╗
║  ║  ║  ║  ║
║  ║  ║  ║  ╠─MissingValue
║  ║  ║  ║  ║ path: a.c.b2
║  ║  ║  ║  ║ Details: optional value, value of type string
║  ║  ║  ║  ▼
║  ║  ║  ║
║  ║  ║  ╠─MissingValue
║  ║  ║  ║ path: a.c.a2
║  ║  ║  ║ Details: optional value, value of type string
║  ║  ║  ▼
║  ║  ▼
║  ║
║  ╠══╦══╗
║  ║  ║  ║
║  ║  ║  ╠─MissingValue
║  ║  ║  ║ path: a.a.b1
║  ║  ║  ║ Details: optional value, value of type string
║  ║  ║  ▼
║  ║  ║
║  ║  ╠─MissingValue
║  ║  ║ path: a.a.a1
║  ║  ║ Details: optional value, value of type string
║  ║  ▼
║  ║
║  ╠══╦══╗
║  ║  ║  ║
║  ║  ║  ╠─MissingValue
║  ║  ║  ║ path: a.a.b2
║  ║  ║  ║ Details: optional value, value of type string
║  ║  ║  ▼
║  ║  ║
║  ║  ╠─MissingValue
║  ║  ║ path: a.a.a2
║  ║  ║ Details: optional value, value of type string
║  ║  ▼
║  ▼
▼)


```

afsalthaj commented:
A few minor comments, and once you make those changes, we are good to merge.

afsalthaj commented:
Or we can consider moving the minor comments in follow up PRs. Let's get this in.
